app_dart/lib/src/request_handlers/push_benchmark_to_center.dart

```diff
@@ -0,0 +1,124 @@
+// Copyright 2020 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+import 'dart:async';
+
+import 'package:appengine/appengine.dart';
+import 'package:gcloud/db.dart';
+import 'package:meta/meta.dart';
+import 'package:metrics_center/flutter.dart' as mc;
+
+import '../datastore/cocoon_config.dart';
+import '../foundation/providers.dart';
+import '../model/appengine/time_series.dart';
+import '../model/appengine/time_series_value.dart';
+import '../request_handling/body.dart';
+import '../request_handling/request_handler.dart';
+import '../service/datastore.dart';
+
+/// Pushes devicelab benchmark results to Metrics Center storage.
+@immutable
+class PushBenchmarkToCenter extends RequestHandler<Body> {
+  PushBenchmarkToCenter(
+    Config config, {
+    @visibleForTesting Logging logger,
+    @visibleForTesting DatastoreService datastoreService,
+  })  : logger = logger ?? Providers.serviceScopeLogger(),
+        datastoreService =
+            datastoreService ?? DatastoreService.defaultProvider(config.db),
+        super(config: config);
+
+  final Logging logger;
+  final DatastoreService datastoreService;
+
+  @override
+  Future<Body> get() async {
+    final int sinceMillis = parseSinceMillis(request.uri.queryParameters);
+    if (sinceMillis == null) {
+      logger.warning('The sinceMillis parameter is missing');
+      return Body.empty;
+    }
+
+    final List<TimeSeriesValue> timeSeriesValues =
+        await readTimeSeriesValue(sinceMillis);
+    final Map<String, TimeSeries> timeSeriesMap = await readTimeSeriesMap();
+
+    final List<mc.MetricPoint> metricPoints =
+        await transform(timeSeriesMap, timeSeriesValues);
+    logger.debug('Transformed ${metricPoints.length} metric points.');
+    logger.debug('The first one is ${metricPoints.first}');
+
+    final mc.FlutterDestination destination =
+        await mc.FlutterDestination.makeFromCredentialsJson(
+            await config.metricsCenterServiceAccountJson);
+    await destination.update(metricPoints);
+    return Body.empty;
+  }
+
+  @visibleForTesting
+  int parseSinceMillis(Map<String, String> queryParameters) {
+    final String s = queryParameters['sinceMillis'] ?? '';
+    return int.tryParse(s);
+  }
+
+  /// Returns a map containing all time series keyed by its id.
+  Future<Map<String, TimeSeries>> readTimeSeriesMap() async {
+    final Map<String, TimeSeries> map = <String, TimeSeries>{};
+    await for (TimeSeries ts in datastoreService.db.query<TimeSeries>().run()) {
+      if (!(ts.id is String)) {
```

Comment:
this condition seems not necessary: https://github.com/flutter/cocoon/blob/master/app_dart/lib/src/model/appengine/time_series.dart#L57